File: python/core/GraphPrimitives.py
from typing import Tuple, NamedTuple, Dict, List, Optional
#from typing import Optional, List, Dict, Any, Type, Callable, TYPE_CHECKING
from collections import  defaultdict

import uuid

# Defining Edge as a simple data structure
# Using NamedTuple for immutability and simple hashability if needed later
class Edge(NamedTuple):
    from_node_id: str
    from_port_name: str
    to_node_id: str
    to_port_name: str
    
    # Optional metadata for debug or visualization, but core logic should rely on first 4 fields
    edge_type: str = "default" # "data", "control"

    def __repr__(self):
        return f"Edge({self.from_node_id}.{self.from_port_name} -> {self.to_node_id}.{self.to_port_name})"


# Base class for Nodes. Will be subclassed by actual node implementations. 
# Contains common properties and methods required for graph traversal.
class GraphNode:
    def __init__(self, 
                 name: str, 
                 type: str, 
                 network_id: str = None
                    ):
        self.name = name
        self.id = uuid.uuid4().hex
        self.uuid = uuid.uuid4().hex  # Unique identifier for the node instance
        self.network_id = network_id

    def __repr__(self):
        return f"GraphNode({self.id})" 

class Graph:
    #all_nodes = {}  # type: Dict[str, 'GraphNode']
    #all_nodes_by_id = {}  # type: Dict[str, 'GraphNode']

    nodes: Dict[str, GraphNode] = {}  # Dictionary of nodes in the net
    edges: List[Edge] = [] # Centralized connection storage (Arena Pattern)

    incoming_edges = defaultdict(list)  # type: Dict[Tuple[str, str], List[Edge]]
    outgoing_edges = defaultdict(list)  # type: Dict[Tuple[str, str], List[Edge]]

    def __init__(self):
        pass
    
     # find a node in all networks by id
    def find_node_by_id(self, uid: str) -> Optional[GraphNode]:
        return self.nodes.get(uid)
    

    def add_edge(self, from_node_id: str, from_port_name: str, to_node_id: str, to_port_name: str):
        # Validation could happen here or in upper layers
        edge = Edge(from_node_id, from_port_name, to_node_id, to_port_name)
        self.edges.append(edge)

        self.incoming_edges[(to_node_id, to_port_name)].append(edge)
        self.outgoing_edges[(from_node_id, from_port_name)].append(edge)
        return edge

    def get_incoming_edges(self, node_id: str, port_name: str) -> List[Edge]:
        return self.incoming_edges.get((node_id, port_name), [])
       
    def get_outgoing_edges(self, node_id: str, port_name: str) -> List[Edge]:
        return self.outgoing_edges.get((node_id, port_name), [])
        


    def add_node(self, node: GraphNode):
        #if self.find_node(node.id):
        #    raise ValueError(f"Node with id '{node.id}' already exists in the global node registry {NodeNetwork.all_nodes.keys()}")
        if self.nodes.get(node.id):
            raise ValueError(f"Node with id '{node.name}' already exists in the network")
       
        

        self.nodes[node.id] = node
        # TODO: why am I not setting the network on the node here?
        # TODO: probably should pass it in as an arg on node create.
        #node.set_network(self) # Inject network context

        print(f"### Graph: Adding node {node.name} to global registry", node.network_id)
        self.nodes[node.id] = node

    
    # TODO: this should be get_node_by_id for clarity
    def get_node_by_id(self, node_id: str) -> Optional[str]:
        return self.nodes.get(node_id)

    # This method looks at nodes LOCAL to this network only
    def get_node_by_name(self, name: str) -> Optional[str]:
        for node in self.nodes.values():
            if node.name == name:
                return node
        return None
    
    def get_node_by_path(self, path: str) -> Optional[GraphNode]:
        #return NodeNetwork.graph.get_node_by_path(path)
        for node in self.nodes.values():
            if node.get_path() == path:
                return node
        return None
    
    def getNode(self, node_id: str) -> Optional[GraphNode]:
        return self.get_node_by_id(node_id)

     # convenience method to get the network node object from the network id
    def getNetwork(self, network_id: Optional[str] = None) -> 'Graph':
        network = self.get_node_by_id(network_id)
        if not network:
            return None
        #assert(network is not None), f"Network with ID '{network_id}' not found"
        assert(network.isNetwork()), f"Node with ID '{network_id}' is not a Network"   
        return network

    # build up the path from a givent node id
    def get_path(self, node_id) -> str:
        
        node = self.get_node_by_id(node_id)
        assert(node is not None), f"Node with ID '{node_id}' not found"

        path_elements = []
        #cur_parent =self.network
        cur_parent = self.getNetwork(node.network_id)
        while cur_parent:
            path_elements.append(cur_parent.name)
            #cur_parent = cur_parent.network
            cur_parent = self.getNetwork(cur_parent.network_id) 
        
        path_elements.reverse()
        full_path = "/" + "/".join(path_elements)
        if node.isNetwork():
            full_path += f"/{node.name}"
        else:
            full_path += f":{node.name}"

        # TODO: fix this hack please.
        if full_path.startswith("//"):
            full_path = full_path[1:]   
        return full_path
    

    def deleteNode(self, name: str):
    
        
        node = self.get_node_by_id(name)
        if not node:
            raise ValueError(f"Node with id '{name}' does not exist in the network")
        

        id  = node.id
        # Arena Pattern: Cleanup connections associated with this node
        self.edges = [e for e in self.edges if e.from_node_id != id and e.to_node_id != id]

        # TODO: deleting a node involves removing all connections to/from it first
        # This cleanup is critical in Rust/TS to prevent orphaned pointers
        del self.nodes[id]
    
    def reset(self):
        self.nodes.clear()
        self.edges.clear()
        self.incoming_edges.clear()
        self.outgoing_edges.clear()
    

File: typescript/src/core/GraphPrimitives.ts
import { v4 as uuidv4 } from 'uuid';
import { IGraphNode, INodePort, INode } from './Interface';

// ─────────────────────────────────────────────
// Edge — immutable connection record (mirrors Python NamedTuple)
// ─────────────────────────────────────────────

export class Edge {
  readonly from_node_id: string;
  readonly from_port_name: string;
  readonly to_node_id: string;
  readonly to_port_name: string;
  readonly edge_type: string;

  constructor(
    fromNodeId: string,
    fromPortName: string,
    toNodeId: string,
    toPortName: string,
    edgeType: string = 'default',
  ) {
    this.from_node_id = fromNodeId;
    this.from_port_name = fromPortName;
    this.to_node_id = toNodeId;
    this.to_port_name = toPortName;
    this.edge_type = edgeType;
  }

  toString(): string {
    return `Edge(${this.from_node_id}.${this.from_port_name} -> ${this.to_node_id}.${this.to_port_name})`;
  }
}

// ─────────────────────────────────────────────
// GraphNode — base class for graph nodes
// ─────────────────────────────────────────────

export class GraphNode extends IGraphNode {
  name: string;
  id: string;
  uuid: string;
  network_id: string | null;

  constructor(name: string, type: string, networkId: string | null = null) {
    super();
    this.name = name;
    this.id = uuidv4().replace(/-/g, '');
    this.uuid = uuidv4().replace(/-/g, '');
    this.network_id = networkId;
  }

  isNetwork(): boolean {
    return false;
  }

  toString(): string {
    return `GraphNode(${this.id})`;
  }
}

// Helper to make a Map key from a (nodeId, portName) tuple
function portKey(nodeId: string, portName: string): string {
  return `${nodeId}::${portName}`;
}

// ─────────────────────────────────────────────
// Graph — arena-pattern graph container
// ─────────────────────────────────────────────

export class Graph {
  nodes: Map<string, IGraphNode>;
  edges: Edge[];
  /** keyed by portKey(to_node_id, to_port_name) */
  private incoming_edges: Map<string, Edge[]>;
  /** keyed by portKey(from_node_id, from_port_name) */
  private outgoing_edges: Map<string, Edge[]>;

  constructor() {
    this.nodes = new Map();
    this.edges = [];
    this.incoming_edges = new Map();
    this.outgoing_edges = new Map();
  }

  /** Find a node in the graph by id */
  find_node_by_id(uid: string): IGraphNode | null {
    return this.nodes.get(uid) ?? null;
  }

  add_edge(
    fromNodeId: string,
    fromPortName: string,
    toNodeId: string,
    toPortName: string,
  ): Edge {
    const edge = new Edge(fromNodeId, fromPortName, toNodeId, toPortName);
    this.edges.push(edge);

    // incoming_edges indexed by destination
    const inKey = portKey(toNodeId, toPortName);
    if (!this.incoming_edges.has(inKey)) this.incoming_edges.set(inKey, []);
    this.incoming_edges.get(inKey)!.push(edge);

    // outgoing_edges indexed by source
    const outKey = portKey(fromNodeId, fromPortName);
    if (!this.outgoing_edges.has(outKey)) this.outgoing_edges.set(outKey, []);
    this.outgoing_edges.get(outKey)!.push(edge);

    // Mirror Python assertion: incoming edges for the FROM port should be < 2
    const fromInKey = portKey(fromNodeId, fromPortName);
    const fromIncoming = this.incoming_edges.get(fromInKey) ?? [];
    console.assert(
      fromIncoming.length < 2,
      'Incoming edge has more than 1 edge',
    );

    return edge;
  }

  get_incoming_edges(nodeId: string, portName: string): Edge[] {
    return this.incoming_edges.get(portKey(nodeId, portName)) ?? [];
  }

  get_outgoing_edges(nodeId: string, portName: string): Edge[] {
    return this.outgoing_edges.get(portKey(nodeId, portName)) ?? [];
  }

  add_node(node: IGraphNode): void {
    if (this.nodes.has(node.id)) {
      throw new Error(`Node with id '${node.name}' already exists in the network`);
    }
    console.log(`### Graph: Adding node ${node.name} to global registry`, node.network_id);
    this.nodes.set(node.id, node);
  }

  /** Alias: mirrors Python Graph.get_node_by_id */
  get_node_by_id(nodeId: string): IGraphNode | null {
    return this.nodes.get(nodeId) ?? null;
  }

  /** Alias: mirrors Python Graph.getNode */
  getNode(nodeId: string): IGraphNode | null {
    return this.get_node_by_id(nodeId);
  }

  get_node_by_name(name: string): IGraphNode | null {
    for (const node of this.nodes.values()) {
      if (node.name === name) return node;
    }
    return null;
  }

  get_node_by_path(path: string): IGraphNode | null {
    for (const node of this.nodes.values()) {
      if (this.get_path(node.id) === path) return node;
    }
    return null;
  }

  getNetwork(networkId: string | null): IGraphNode | null {
    if (!networkId) return null;
    const network = this.get_node_by_id(networkId);
    if (!network) return null;
    if (!network.isNetwork()) {
      throw new Error(`Node with ID '${networkId}' is not a Network`);
    }
    return network;
  }

  get_path(nodeId: string): string {
    const node = this.get_node_by_id(nodeId);
    if (!node) throw new Error(`Node with ID '${nodeId}' not found`);

    const pathElements: string[] = [];
    let curParent = this.getNetwork(node.network_id);
    while (curParent) {
      pathElements.push(curParent.name);
      curParent = this.getNetwork(curParent.network_id);
    }

    pathElements.reverse();
    let fullPath = '/' + pathElements.join('/');
    if (node.isNetwork()) {
      fullPath += `/${node.name}`;
    } else {
      fullPath += `:${node.name}`;
    }

    // Fix double-slash at root
    if (fullPath.startsWith('//')) {
      fullPath = fullPath.slice(1);
    }
    return fullPath;
  }

  deleteNode(nodeId: string): void {
    const node = this.get_node_by_id(nodeId);
    if (!node) {
      throw new Error(`Node with id '${nodeId}' does not exist in the network`);
    }

    const id = node.id;

    // Arena Pattern: Cleanup connections associated with this node
    this.edges = this.edges.filter(
      (e) => e.from_node_id !== id && e.to_node_id !== id,
    );

    // Rebuild incoming/outgoing maps after deletion
    this._rebuildEdgeMaps();

    this.nodes.delete(id);
  }

  private _rebuildEdgeMaps(): void {
    this.incoming_edges.clear();
    this.outgoing_edges.clear();
    for (const edge of this.edges) {
      const inKey = portKey(edge.to_node_id, edge.to_port_name);
      if (!this.incoming_edges.has(inKey)) this.incoming_edges.set(inKey, []);
      this.incoming_edges.get(inKey)!.push(edge);

      const outKey = portKey(edge.from_node_id, edge.from_port_name);
      if (!this.outgoing_edges.has(outKey)) this.outgoing_edges.set(outKey, []);
      this.outgoing_edges.get(outKey)!.push(edge);
    }
  }

  reset(): void {
    this.nodes.clear();
    this.edges = [];
    this.incoming_edges.clear();
    this.outgoing_edges.clear();
  }

  get_downstream_ports(srcPort: INodePort, includeIoPorts: boolean = false): INodePort[] {
    const downstreamPorts: INodePort[] = [];
    const outgoingEdges = this.get_outgoing_edges((srcPort as any).node_id, (srcPort as any).port_name);

    for (const edge of outgoingEdges) {
      const destNode = this.get_node_by_id(edge.to_node_id) as any;
      if (!destNode) continue;

      let destPort: INodePort | null = destNode.inputs?.[edge.to_port_name] ?? null;
      if (!destPort) destPort = destNode.outputs?.[edge.to_port_name] ?? null;
      if (!destPort) continue;

      if ((destPort as any).isInputOutputPort()) {
        if (includeIoPorts) downstreamPorts.push(destPort as INodePort);
        downstreamPorts.push(
          ...this.get_downstream_ports(destPort as INodePort, includeIoPorts),
        );
      } else {
        downstreamPorts.push(destPort as INodePort);
      }
    }

    return downstreamPorts;
  }

  get_upstream_ports(port: INodePort, includeIoPorts: boolean = false): INodePort[] {
    const upstreamPorts: INodePort[] = [];
    const incomingEdges = this.get_incoming_edges((port as any).node_id, (port as any).port_name);

    for (const edge of incomingEdges) {
      const srcNode = this.get_node_by_id(edge.from_node_id) as any;
      if (!srcNode) continue;

      let srcPort: INodePort | null = srcNode.outputs?.[edge.from_port_name] ?? null;
      if (!srcPort) srcPort = srcNode.inputs?.[edge.from_port_name] ?? null;
      if (!srcPort) continue;

      if ((srcPort as any).isInputOutputPort()) {
        if (includeIoPorts) upstreamPorts.push(srcPort as INodePort);
        upstreamPorts.push(...this.get_upstream_ports(srcPort as INodePort));
      } else {
        upstreamPorts.push(srcPort as INodePort);
      }
    }

    return upstreamPorts;
  }

  get_upstream_nodes(port: INodePort): IGraphNode[] {
    const upstreamNodes: IGraphNode[] = [];
    const incomingEdges = this.get_incoming_edges(
      (port as any).node_id,
      (port as any).port_name,
    );
    for (const edge of incomingEdges) {
      const srcNode = this.get_node_by_id(edge.from_node_id);
      if (srcNode && !upstreamNodes.includes(srcNode)) {
        upstreamNodes.push(srcNode);
      }
    }
    return upstreamNodes;
  }

  get_downstream_nodes(port: INodePort): IGraphNode[] {
    const downstreamNodes: IGraphNode[] = [];
    const outgoingEdges = this.get_outgoing_edges(
      (port as any).node_id,
      (port as any).port_name,
    );
    for (const edge of outgoingEdges) {
      const destNode = this.get_node_by_id(edge.to_node_id);
      if (destNode && !downstreamNodes.includes(destNode)) {
        downstreamNodes.push(destNode);
      }
    }
    return downstreamNodes;
  }
}
